### PR TITLE
Development PGD

### LINKED ISSUE #714: Change Implementation of Random Restarts in PGD
In it's current implementation PGD with N random restarts runs a batch N times though the PGD attack and picks the batch with the worst performance with the accuracy being computed averaged on a batch basis.

However, this can give overly optimistic scores of model performance. As a simple example, consider two random restarts and a batch size of 10. The first restart the first 5 examples are classified correctly. The second restart the last 5 are classified correctly. In both cases the accuracy is 50% averaged across the batch. However, on a per sample basis (ie the worst score achieved for a given example) the score is 0%.

As an improvement we can change how the accuracy is computed so that is conducted over each sample.

### PR DESCRIPTION
**Pull request by GiulioZizzo**
# Description

We are changing how random restarts are implemented for PGD to select the worst case per example to return, rather then the worst case averaged across the whole datasets.

Fixes #714

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- Accuracy of random restarts is now more effective when testing against various locally trained classifiers under tensorflow and sklearn.

**Test Configuration**:
- OS:  RHEL
- Python version: 3.6 and 3.7
- ART version or commit number: e1dbdb9320c50bd06b0ff1c061f5022d830f2ff4
- TensorFlow / Keras / PyTorch / MXNet version: tensorflow==2.3.1, Keras==2.4.3, torch==1.6.0+cu92, mxnet==1.6.0

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py [hunk 1]
```diff
@@ -259,43 +255,62 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
...
+
+            for batch_id in range(int(np.ceil(x.shape[0] / float(self.batch_size)))):
+                for rand_init_num in trange(
+                    max(1, self.num_random_init), desc="PGD - Random Initializations", disable=not self.verbose
+                ):
+                    batch_index_1, batch_index_2 = batch_id * self.batch_size, (batch_id + 1) * self.batch_size
+                    batch_index_2 = min(batch_index_2, x.shape[0])
+                    batch = x[batch_index_1:batch_index_2]
+                    batch_labels = targets[batch_index_1:batch_index_2]
+                    mask_batch = mask
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py:268]: Can we combine this line with the code in line 270?

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py [hunk 2]
```diff
@@ -259,43 +255,62 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
...
+
+                    if mask is not None:
+                        if len(mask.shape) == len(x.shape):
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py:271]: Can we combine these two if conditions into one?
Should we raise an exception if len(mask.shape) != len(x.shape)?

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py [hunk 3]
```diff
@@ -226,17 +234,26 @@ def _generate_batch(
 
         :param x: An array with the original inputs.
         :param targets: Target values (class labels) one-hot-encoded of shape `(nb_samples, nb_classes)`.
+        :param mask: An array with a mask to be applied to the adversarial perturbations. Shape needs to be
+                     broadcastable to the shape of x. Any features for which the mask is zero will not be adversarially
+                     perturbed.
+        :param eps: Maximum perturbation that the attacker can introduce.
+        :param eps_step: Attack step size (input variation) at each iteration.
```
Comment (beat-buesser) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:241]: "param mask" is now twice in the docstring, let's keep the one in line 237 which is consistent with other attacks

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @GiulioZizzo Thank you very much for your first pull request to ART! This pull request will improve the effectiveness of random restarts in all implementations of the key attack ProjectedGradientDescent!

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py (hunk 1 above)
Comment (GiulioZizzo) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py:268]: Hi Beat and thanks for the comments! This is following the coding pattern used in fast_gradient.py so perhaps we can keep it as it is for consistency between code.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py (hunk 2 above)
Comment (GiulioZizzo) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py:271]: Hi Beat and thanks for the comments! This is following the coding pattern used in fast_gradient.py so perhaps we can keep it as it is for consistency between code.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py (hunk 3 above)
Comment (GiulioZizzo) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:241]: Hi- the redundant docstring has been removed now!

**Review by beat-buesser — APPROVED** [review]
